qubes/tests/__init__.py

```diff
@@ -1127,6 +1127,50 @@ def list_templates():
             _templates = ()
     return _templates
 
+def create_testcases_for_templates(name, *bases, globals, **kwds):
+    '''Do-it-all helper for generating per-template tests via load_tests proto
+
+    This does several things:
+        - creates per-template classes
+        - adds them to module's :py:func:`globals`
```

Comment:
What about adding them to module object instead (`sys.modules[...]`)? As you've done for `extra` tests. IMHO this would be slightly less hacky... You can then pass simply a module name (to avoid looking at stack, as you've explained below).

Reply:
This is the same:
```python
import sys
assert globals() is sys.modules[__name__].__dict__
```
And explicit is better than implicit™ [PEP-20]. The reason for `extra` is that there is no other viable place to add this to (endpoint loading is purposefully constrained).

Reply:
Still, IMO better use sys.modules, makes code less magic (easier to understand for non-python-wizard) and more flexible. Right now it greatly depends on the _place_ where `globals()` is called, using `sys.modules` would avoid this.